### ISSUE TITLE
Farmland stays when jumping

### ISSUE THREAD

**Issue by cflep** [cuberite/cuberite#5069]
Client version: 1.11.2
Server OS: Linux
Cuberite Commit id: Cuberite local build (id: Unknown)

### Expected behavior
When jumping on farmland, the block should be reset to dirt.

### Actual behavior
Nothing happens

### Steps to reproduce the behavior
Jump on farmland

**Comment by cflep** [comment]
I'll try to implement the right behaviour, but it might take a while, I'm not that experienced in C++

**Comment by dImrich** [comment]
We need to add something like vanilla onFallenUpon block event to do this.

**Comment by cflep** [comment]
I tried adding a few lines to /src/Blocks/BlockFarmland.h and just when I thought I'd corrected all the errors I got the following:

`In file included from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../ChunkMap.h:10,
                 from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../World.h:9,
                 from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/Player.h:7,
                 from /home/user/Dokumente/github/cuberite/src/Blocks/BlockDoor.h:6,
                 from /home/user/Dokumente/github/cuberite/src/Blocks/BlockDoor.cpp:3,
                 from /home/user/Dokumente/github/cuberite/CMakeFiles/Cuberite.dir/Unity/unity_13_cxx.cxx:3:
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h: In instantiation of ‘static Ret cFunctionRef<Ret(Args ...)>::ObjectFunctionCaller(void*, Args ...) [with ObjectType = cBlockFarmlandHandler::OnPlayerFall(const cChunk&, Vector3i, cChunkInterface&, Vector3i)::<lambda(cEntity&)>; Ret = bool; Args = {cEntity&}]’:
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h:51:18:   required from ‘cFunctionRef<Ret(Args ...)>::cFunctionRef(FunctionObject&&) [with FunctionObject = cBlockFarmlandHandler::OnPlayerFall(const cChunk&, Vector3i, cChunkInterface&, Vector3i)::<lambda(cEntity&)>; typename std::enable_if<(! std::is_same<typename std::decay<_Functor>::type, cFunctionRef<Ret(Args ...)> >::value), int>::type <anonymous> = 0; Ret = bool; Args = {cEntity&}]’
/home/user/Dokumente/github/cuberite/src/Blocks/BlockFarmland.h:178:4:   required from here
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h:71:16: error: void value not ignored as it ought to be
   71 |   return Object(std::forward<Args>(a_Args)...);
      |          ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/Cuberite.dir/build.make:556: CMakeFiles/Cuberite.dir/Unity/unity_13_cxx.cxx.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:866: CMakeFiles/Cuberite.dir/all] Fehler 2
make: *** [Makefile:130: all] Fehler 2

`
I don't know what the problem is...
Again, I'm sorry I'm not that used to C++, usually I use Python :')

**Comment by cflep** [comment]
I tried adding a few lines to /src/Blocks/BlockFarmland.h and just when I thought I'd corrected all the errors I got the following:
```
In file included from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../ChunkMap.h:10,
from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../World.h:9,
from /home/user/Dokumente/github/cuberite/src/Blocks/../Entities/Player.h:7,
from /home/user/Dokumente/github/cuberite/src/Blocks/BlockDoor.h:6,
from /home/user/Dokumente/github/cuberite/src/Blocks/BlockDoor.cpp:3,
from /home/user/Dokumente/github/cuberite/CMakeFiles/Cuberite.dir/Unity/unity_13_cxx.cxx:3:
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h: In instantiation of ‘static Ret cFunctionRef<Ret(Args ...)>::ObjectFunctionCaller(void*, Args ...) [with ObjectType = cBlockFarmlandHandler::OnPlayerFall(const cChunk&, Vector3i, cChunkInterface&, Vector3i)::<lambda(cEntity&)>; Ret = bool; Args = {cEntity&}]’:
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h:51:18: required from ‘cFunctionRef<Ret(Args ...)>::cFunctionRef(FunctionObject&&) [with FunctionObject = cBlockFarmlandHandler::OnPlayerFall(const cChunk&, Vector3i, cChunkInterface&, Vector3i)::<lambda(cEntity&)>; typename std::enable_if<(! std::is_same<typename std::decay<_Functor>::type, cFunctionRef<Ret(Args ...)> >::value), int>::type = 0; Ret = bool; Args = {cEntity&}]’
/home/user/Dokumente/github/cuberite/src/Blocks/BlockFarmland.h:178:4: required from here
/home/user/Dokumente/github/cuberite/src/Blocks/../Entities/../FunctionRef.h:71:16: error: void value not ignored as it ought to be
71 | return Object(std::forward(a_Args)...);
| ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/Cuberite.dir/build.make:556: CMakeFiles/Cuberite.dir/Unity/unity_13_cxx.cxx.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:866: CMakeFiles/Cuberite.dir/all] Fehler 2
make: *** [Makefile:130: all] Fehler 2

```
I don't know what's causing the issue...
Again, I'm sorry I'm not that used to C++, usually I use Python :')

**Comment by 12xx12** [comment]
It seams that you try to use a callback

what is your branch called? would like to check the code out.

**Comment by cflep** [comment]
It's FarmlandJumping in my fork

**Comment by 12xx12** [comment]
oh - is it published?

**Comment by cflep** [comment]
I also messed up some other files, but that shouldn't matter

**Comment by cflep** [comment]
Should be
[https://github.com/cflep/cuberite/tree/FarmlandJumping](url)

**Comment by 12xx12** [comment]
found it

**Comment by 12xx12** [comment]
you have to return something in that metod

**Comment by cflep** [comment]
Got it, thanks :+1:

**Comment by 12xx12** [comment]
Let me rephrase - you have to return false to continue iterating
if you return true the iteration will cancel

**Comment by koutsie** [comment]
In my opinion you should make FarmlandJumping an option in world.ini :flushed:

**Comment by DarkoGNU** [comment]
Hey, @cflep seems to have abandoned this issue, so I'll try to implement farmland trampling

**Comment by DarkoGNU** [comment]
I measured how many falls from different heights (measured in blocks) are required to trample 50 farmlands in order to study the vanilla trampling behavior.

Testing procedure:
- version - vanilla 1.12.2
- 1/8 of a block - one snow
- farmlands without crops
- manual testing - tedious jumping on farmlands :)

![image](https://user-images.githubusercontent.com/42816979/161406045-87f738c0-f8e3-4a3a-b5e5-701717da23fa.png)

Results:

6/8: no trample after 50 falls (0% chance)
1 and 6/8: 50 falls (100% chance)

7/8 - 1 and 1/8:
7/8 of a block: 201 falls (min: 1, max: 18, chance: 25%)
1 block: 205 falls (min: 1, max: 26, chance: 24%)
1 and 1/8: 204 falls (min: 1, max: 11, chance: 25%)
Average chance: 25% (150 ÷ 610)

1 and 2/8: 69 falls (min: 1, max: 3, chance: 74%)
1 and 3/8: 75 falls (min: 1, max: 3, chance: 66%)
1 and 4/8: 87 falls (min: 1, max: 6, chance: 57%)
1 and 5/8: 63 falls (min: 1, max: 3, chance: 79%)
Average chance: 68% (200 ÷ 294)

Important exceptions:
- mobs smaller than 0.512 cubic block cannot destroy farmland (source: Minecraft Wiki)
- an enderman teleporting directly on top of a farmland block will trample it (source: Minecraft Wiki)
- a ravager will always trample the farmland (source: Minecraft Wiki)

So, it looks like my code will need to work like this:
- fall from [0, 6/8] - never trample
- fall from (6/8, 1 and 1/8] - 25% chance of trample
- fall from (1 and 1/8, 1 and 5/8] - 66% chance of trample (rounding a little bit here)
- fall from (1 and 5/8, ∞) - always trample

Of course, if someone looks at Microsoft's source code, it might turn out that these chances are different. My sample size was pretty low. But it's better to base my implementation on this than just guessing. Is there any good way to acquire information from Minecraft's source code tho? Would e.g. asking someone not involved in Cuberite to decompile Minecraft and let me know how is the chance of trampling calculated be OK? Or is it better to stick to experiments?

Anyway, now I can begin coding. Tomorrow... I mean - after I wake up (it's 2 AM in my time zone right now). It can take some time, I'm kinda contributing code to a big project for the first time
